### ISSUE TITLE
Exception handling: device initialisation after recovery from exception

### ISSUE THREAD

**Issue by mhier** [ChimeraTK/ApplicationCore#46]
This ticket is a child of #10.

Background: If a device is recovered after an exception, it might need to be reinitialised (e.g. because it was power cycled). Currently, device initialisation is completely up to the application. This needs to be changed so devices can be automatically reinitialised after recovery.

Task:

- Extend the interface of DeviceModule so the application can define a callback function to be executed when the device needs to be initialised.
- The function shall have one argument of the type ChimeraTK::Device (the device to be initialised) and a no return value. The function is executed inside the DeviceModule thread.
- The function can be either specified in the DeviceModule constructor or through a later call to a DeviceModule member function named DeviceModule::addInitialisationHandler(). Specifying the function is optional (a device might not require any initialisation).
- By calling DeviceModule::addInitialisationHandler() multiple times it is possible to add multiple initialisation handlers. The handlers are called in the order they were added (the one added in the DeviceModule constructor comes first).
- The handlers shall called in two situations (by the DeviceModule):
  - The device has just been successfully opened for the first time during the initialisation phase
  - The device has been successfully opened again after recovery from an exception state
- The handler may throw ChimeraTK::runtime_error. This exception should be treated in the same way as it were reported by DeviceModule::reportException(), i.e. the device goes again into the exception state (or does not even leave it) and the exception what() string is shown to the control system. This allows the function to also report errors on an application level by throwing the exception explicitly, e.g. if the hardware state is not like expected. Please mention this in the documentation.
- If a handler has thrown an exception, the remaining handlers are not called (until the next attempt to recover the device).

Note: multiple handlers are required for the implementation of #47 (as it will add initialisation handlers to devices which might already have a user-defined handler).

DoD:

- [x] Implementation
- [x] Tests:
  - Handler gets called in initialisation phase of the application.
  - Handler gets called after exception.
  - Correct behaviour if function throws ChimeraTK::runtime_error.
  - Correct behaviour with multiple handlers.
- [x] Documentation

**Comment by killenb** [comment]
This task is blocked by bug #53. I have to postpone it until this is solved.

**Comment by killenb** [comment]
Solved in aaf2beb083cfb80d4f57231b3a1173f930106d0a, toghether with #53
